posthog/migrations/0173_should_update_person_props_function.py

```diff
@@ -0,0 +1,38 @@
+# Generated by Django 3.1.12 on 2021-10-05 13:06
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("posthog", "0172_person_properties_last_operation"),
+    ]
+
+    operations = [
+        migrations.RunSQL(
+            """
+        CREATE OR REPLACE FUNCTION should_update_person_prop(
+            in person_id int, 
+            in prop_key text, 
+            in _timestamp text, 
+            in operation text, 
+            out should_update bool
+        )
+        AS $$ 
+            SELECT NOT property_exists OR
+            (operation='set' AND is_timestamp_newer) OR
+            (last_operation='set_once' AND NOT is_timestamp_newer)
+            FROM (
+                SELECT 
+                    properties->prop_key IS NOT NULL as property_exists, 
+                    _timestamp > COALESCE(properties_last_updated_at ->> prop_key, '0') as is_timestamp_newer, 
+                    COALESCE(properties_last_operation ->> prop_key, '') as last_operation 
```

Comment:
1. Let's be super explicit here on what we assume if the props (last_updated_at & last_operation) didn't exist before, specifically if `properties_last_operation` isn't set we should assume it was `$set`, and the `last_updated_at` we can assume beginning of time, for which '0' works fine.
2. The case where the timestamps are equal, we should treat it as newer, because we're processing this event later, hence we can assume it was later (in the table too, update it to say s/after/equal or after/ ). An example case, where this could happen is someone's app is setting timestamps & they set them the same for the batch, but within the batch there is still order.

Maybe add tests also for `test_with_equal_timestamp`, `test_without_last_operation`, `test_without_last_timestamp_without_last_operation` (based on the order we did things we can't have a case, where last_operation exists & last_timestamp doesn't, so no need to test that).

```suggestion
                    _timestamp >= COALESCE(properties_last_updated_at ->> prop_key, '0') as is_timestamp_newer, 
                    COALESCE(properties_last_operation ->> prop_key, 'set') as last_operation 
```

Reply:
Regarding the coalesce for `properties_last_operation`, I'm happy to set the default to `set`. It doesn't really matter though, given we're only checking if it is `set_once` or not, so any value (empty string included) should be enough.

Regarding timestamps being equal, I'm not sold. I don't think we can reasonably infer out of 2 events with the same exact timestamp which one was sent earlier. There are a lot of opportunities in our pipeline for the order to flip. Therefore, I'd think `>` is better, because it prevents us from doing an additional update (the query to update props won't be super cheap). Given someone sends us events with the same exact timestamp, I think it's fair to say we'll process the first update and that's it.

Reply:
`test_without_last_operation` and `test_without_last_timestamp` could be added but they are unnecessary. 

These will always be there if the prop exists, so the test that checks for the prop not existing is all we should really need here.

Although as I'm writing this I realize that would imply the `COALESCE` isn't needed but we probably want to have that there anyway.

Reply:
So actually there's a tradeoff here. With the current implementation, this actually looks like this:

```
# 11| set      | yes           | equal                      | set         | no
# 12| set      | yes           | equal                      | set_once    | yes
# 13| set_once | yes           | equal                      | set         | no
# 14| set_once | yes           | equal                      | set_once    | yes
```

Reason being we do `NOT is_timestamp_newer`. To achieve this effect we'd need `is_timestamp_equal` as well. This is a bit more complexity and harms readability a bit, so I've left as is.

You can push back though and we can introduce `is_timestamp_equal` too to make this consistent. We'd be doing more work to figure this out to cut down on a few updates, so I think it's actually fine as is.